### ISSUE TITLE
Emit metric for unconsumed outputs

### ISSUE THREAD

**Issue by jszwedko** [vectordotdev/vector#11940]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Use Cases

A user is interested in a metric for the number of events output by a `route` transform that are not consumed by another component. That is, for a named output that is not an input.

I think we could take this further and just publish a metric of events discarded by any component outputs that aren't wired up.

### Proposal

Publish a metric (maybe re-use `component_discarded_events_total`?) for events written to an output that are not consumed by another component.

### References

Supersedes https://github.com/vectordotdev/vector/issues/11238

### Version

vector 0.20.0 (x86_64-apple-darwin 2a706a3 2022-02-10)

**Comment by JeanMertz** [comment]
Alternatively, we could _enforce_ the requirement of each route feeding into an existing component. That way, no unexpected typo errors could creep in. You can then feed anything you don't need any more into a `blackhole` sink, which _does_ emit metrics.

**Comment by jszwedko** [comment]
> Alternatively, we could _enforce_ the requirement of each route feeding into an existing component. That way, no unexpected typo errors could creep in. You can then feed anything you don't need any more into a `blackhole` sink, which _does_ emit metrics.
>
> Taking this "one step further", as you also mentioned above would be to apply that same constraint to all components, and requiring any topology to always end with a sink.

Yeah, I have toyed with the idea of a "strict mode" that would enforce this among other things.

> And if we wanted to make such a change backward compatible, we could automatically attach a `blackhole` sink to any unconsumed non-sink component, and potentially emit a warning at boot to alert the operator of this fact (although I think I remember seeing such a warning already being emitted, we just don't auto-attach a blackhole sink when such a situation occurs).

Hmm, yeah, that's an interesting idea. It makes monitoring this metric requested here a little indirect, but is tractable.
